mscore/scoreview.cpp

```diff
@@ -5263,6 +5266,50 @@ qDebug("midiNoteReceived %d chord %d", pitch, chord);
       }
 
 //---------------------------------------------------------
+//   realtimeAdvance
+//---------------------------------------------------------
+
+void ScoreView::realtimeAdvance(bool allowRests)
+      {
+      if (!noteEntryMode() || (!allowRests && score()->activeMidiPitches()->empty()))
+            return;
+      InputState& is = score()->inputState();
+      switch (is.noteEntryMethod()) {
+            case NoteEntryMethod::REALTIME_MANUAL:
+                  cmdRealtimeAdvance();
+                  break;
+            case NoteEntryMethod::REALTIME_AUTO:
+                  { // curly braces needed to protect intitialisations. See: http://stackoverflow.com/a/5685578
+                  static bool loop = false;
+                  if (loop) {
+                        // another loop is already running
+                        if (allowRests)
+                              loop = false; // terminate the other loop
+                        return; // exit function
+                        }
+                  QTime timer;
+                  timer.start();
+                  loop = true;
+                  const int realtimeDelay = preferences.realtimeDelay;
+                  for (;;) {
+                        int delay = realtimeDelay - timer.elapsed(); // account for non-zero execution time of operations inside loop
+                        if (delay > 0)
+                              guiSafeSleep(delay);
+                        timer.start();
+                        if (!loop || !noteEntryMode() || !is.usingNoteEntryMethod(NoteEntryMethod::REALTIME_AUTO)
+                                    || (!allowRests && score()->activeMidiPitches()->empty()))
+                              break;
+                        cmdRealtimeAdvance();
+                        }
```

Comment:
@lasconic, here is the code that should stop the score advancing during auto mode if the advancing was triggered by pressing a note. It works fine on Linux - any idea why it doesn't work on a mac?

```c++
if (!allowRests && score()->activeMidiPitches()->empty())
      break;
```

`allowRests` is a boolean that is passed to the function [here](https://github.com/musescore/MuseScore/pull/2734/commits/4c08ada877ff656d3a55ce716667382d735daabf#diff-39acd419478f3f0d33e494c8b37942b0R3405). `loop` is a static boolean that is declared [here](https://github.com/musescore/MuseScore/pull/2734/commits/4c08ada877ff656d3a55ce716667382d735daabf#diff-39acd419478f3f0d33e494c8b37942b0R5283) and set [here](https://github.com/musescore/MuseScore/pull/2734/commits/4c08ada877ff656d3a55ce716667382d735daabf#diff-39acd419478f3f0d33e494c8b37942b0R5292).